tests/test_preprocessing.py

```diff
@@ -337,6 +339,39 @@ def test_regress_out_reproducible():
     np.testing.assert_allclose(adata.X, tester)
 
 
+def test_regress_out_reproducible_category():
+    adata = sc.datasets.pbmc68k_reduced()
+    adata = adata.raw.to_adata()[:200, :200].copy()
+    sc.pp.regress_out(adata, keys=["bulk_labels"])
+    # This file was generated from the original implementation in version 1.10.3
+    # Now we compare new implementation with the old one
+    tester = np.load(DATA_PATH / "regress_test_small_cat.npy")
+    np.testing.assert_array_almost_equal(adata.X, tester)
+
+
+def test_regressor_categorical():
+    adata = sc.datasets.pbmc68k_reduced()
+    adata = adata.raw.to_adata()[:200, :200]
+    X_org = adata.X.copy().astype(np.float64)
+    keys = ["bulk_labels"]
+    # Create org regressors
+    regressors = np.zeros(X_org.shape, dtype=X_org.dtype)
+    X = _to_dense(X_org, order="F") if issparse(X_org) else X_org
+    for category in adata.obs[keys[0]].cat.categories:
+        mask = (category == adata.obs[keys[0]]).values
+        for ix, x in enumerate(X.T):
+            regressors[mask, ix] = x[mask].mean()
+
+    # Create new regressors
+    cats = np.int64(len(adata.obs[keys[0]].cat.categories))
+    filters = adata.obs[keys[0]].cat.codes.to_numpy()
+    cats = cats.astype(filters.dtype)
+    X = _to_dense(X_org, order="F") if issparse(X_org) else X_org
```

Comment:
This test is not parameterized so `X_org` is either sparse or not, no? So this line could be simplified or the test should be parameterized to handle both cases